core/unix/signal.c

```diff
@@ -6457,6 +6448,21 @@ os_forge_exception(app_pc target_pc, dr_exception_type_t type)
         sig = SIGSEGV;
         break;
     }
+    dr_forge_signal(target_pc, sig);
+}
+
+void
+dr_forge_signal(app_pc target_pc, int sig) {
+    dcontext_t *dcontext = get_thread_private_dcontext();
```

Comment:
The code below uses the mcontext stored in the dcontext.  That is not the current application state when in a clean call (the state is on the dstack then).  Plus, the user may want to set more than just the pc.  I think the API should take in the `dr_mcontext_t` to use.

Reply:
The code uses `get_mcontext(dcontext)` which returns `priv_mcontext_t`, not `dr_mcontext_t`. Also, we cannot support both `dcontext_t *dcontext` and `dr_mcontext_t* mc` (or `priv_mcontext_t* mc`) as input parameters since we would have to deal with all sorts of combinations of those two (most of which would probably be meaningless).
Did you mean that API should take in `dcontext_t *dcontext`?

Reply:
It should be like the other API routines: the exposed interface function should take in the public type dr_mcontext_t and convert it to the private type priv_mcontext_t before passing to an internal function.

Reply:
Thank you for the explanation about `dr_mcontext_t` vs `priv_mcontext_t`. I can see how this should work. Originally, the `os_forge_exception` internal function obtained `dcontext_t` from `get_thread_private_dcontext()` and then used it to get `priv_mcontext_t`. If we expose `dr_mcontext_t` as a separate parameter, should we also expose `dcontext_t` as a separate one?

Addition of `dr_mcontext_t` to the new API function results in more changes to the existing internal function `os_forge_exception`. It becomes inconvenient to use the API function from here (we would need to call `dr_get_mcontext()` to obtain public mcontext which will later be converted into private mcontext).

To satisfy both requirements I should add another internal function which will take `priv_mcontext_t` and will be used from both `os_forge_exception` and `dr_forge_signal`.